src/http2.rs

```diff
@@ -0,0 +1,271 @@
+    /// A thin wrapper around `SimpleClient::request`.
+    pub fn request(&mut self, method: &[u8], path: &[u8], extras: &[(Vec<u8>, Vec<u8>)])
+            -> Http2Result<StreamId> {
+        match self {
+            &mut Http2Client::Http(ref mut client) => client.request(method, path, extras),
+            &mut Http2Client::Https(ref mut client) => client.request(method, path, extras),
+        }
+    }
+
+    /// Read the response to the given stream_id. It will block the calling thread
+    /// until either the response is received or the connection errors.
+    /// A thin wrapper around `SimpleClient::get_response`.
+    pub fn get_response(&mut self, stream_id: StreamId) -> Http2Result<RawHttp2Response> {
+        match self {
+            &mut Http2Client::Http(ref mut client) => client.get_response(stream_id),
+            &mut Http2Client::Https(ref mut client) => client.get_response(stream_id),
+        }
+    }
+}
+
+/// A struct that creates new HTTP/2 requests, such that they can be used in
+/// `hyper`'s `Client` implementation.
+pub struct Http2RequestFactory;
+
+impl Http2RequestFactory {
+    /// Returns a new `Http2Client` that can be used by a request.
+    fn get_client(&mut self, url: &Url) -> Http2Client {
+        // TODO Return real errors without just unwrapping willy-nilly
+        Http2Client::new(&url.scheme, url.domain().unwrap())
+    }
+}
+
+impl HttpRequestFactory for Http2RequestFactory {
+    type RequestType = Http2Request<Fresh>;
+
+    fn new() -> Http2RequestFactory { Http2RequestFactory }
+
+    fn get_fresh_request(&mut self, template: &RequestTemplate) -> HttpResult<Http2Request<Fresh>> {
+        Ok(Http2Request {
+            client: self.get_client(&template.url),
+            headers: Headers::new(),
+            stream_id: None,
+            method: template.method.clone(),
+            url: template.url.clone(),
+            _marker: PhantomData,
+        })
+    }
+
+    fn set_ssl_verifier(&mut self, verifier: ContextVerifier) {
+        // TODO Don't ignore this...
+    }
+}
+
+/// A struct representing an HTTP/2-based request. Satisfies the same interface
+/// that `hyper::client::request::Request` does and allows for a similar
+/// transformation of `Fresh -> Streaming -> Response`
+pub struct Http2Request<W> {
+    client: Http2Client,
+    headers: Headers,
+    stream_id: Option<StreamId>,
+    method: Method,
+    url: Url,
+    _marker: PhantomData<W>,
+}
+
+impl<W> HttpRequest for Http2Request<W> {
+    fn headers(&self) -> &Headers { &self.headers }
+    fn method(&self) -> Method { self.method.clone() }
+}
+
+impl Http2Request<Fresh> {
+    fn start(mut self) -> HttpResult<Http2Request<Streaming>> {
+        // Prepare the request metadata so that it fits into `solicit`'s API.
+        // It basically just adapts the method and path into byte slices.
+        let mut method_bytes: Vec<u8> = Vec::new();
+        write!(&mut method_bytes, "{}", self.method);
```

Comment:
How about we do something similar to what you suggested with using `into_bytes` for the path.

Namely, implement `Borrow<str>` for `Method` and then just pass that (`as_bytes`) to the `solicit` API that expects a `&[u8]`. It would avoid an extra allocation here...